### PR TITLE
Additional signaling

### PR DESCRIPTION
**Pull request by erikwijmans**
In a situation with many consumers and one batched producer, the system can stall since the batched producer will produce enough work for many consumers but its single call to `pthread_cond_signal(&q->not_empty)` is only guaranteed to wake up one consumer.  This PR adds a counter to the number of consumers waiting and has consumers wake each other up if no producer will be woken up and there are still things to be read.

A similar issue exists in the case of many producers and one batched consumer.

### CODE REVIEW

**Review by alex-petrenko — APPROVED** [review]
@erikwijmans thank you for this! Very clever solution.

I remember I though about this problem before... Back then there was no put_many(), but get_many() can cause the same problem.
Since get_many() will only wake a single producer.

Before I merge, I have a couple of questions:
1) Did you try just using broadcast instead of signal? I guess this will wake up too many processes, so potentially bad for performance?
2) Why producers or consumer have to wake each other up? The way it is implemented now, they can only wake up sequentially, one-by-one. I guess it is okay since they need the mutex anyway?
Basically, why i.e. get_many() process can't just signal n producer processes? You think it'll be worse for performance?

**Comment by erikwijmans** [comment]
I didn't look at the performance of a broadcast.  It will certainly hurt performance from an academic standpoint (1 lock grab is always better than n) but IDK if that is actually measurable.
